src/ess/reduce/nexus/workflow.py

```diff
@@ -685,16 +688,34 @@ def GenericNeXusWorkflow(
     wf[DetectorBankSizes] = DetectorBankSizes({})
     wf[PreopenNeXusFile] = PreopenNeXusFile(False)
 
-    g = wf.underlying_graph
-    ancestors = set()
-    # DetectorData and MonitorData are the "final" outputs, so finding and removing all
-    # their ancestors is what we need to strip unused run and monitor types.
-    for rt in run_types or ():
-        ancestors |= nx.ancestors(g, DetectorData[rt])
-        ancestors.add(DetectorData[rt])
-        for mt in monitor_types or ():
-            ancestors |= nx.ancestors(g, MonitorData[rt, mt])
-            ancestors.add(MonitorData[rt, mt])
-    if run_types is not None:
-        g.remove_nodes_from(set(g.nodes) - ancestors)
+    if run_types is not None or monitor_types is not None:
+        _prune_type_vars(wf, run_types=run_types, monitor_types=monitor_types)
+
     return wf
+
+
+def _prune_type_vars(
+    workflow: sciline.Pipeline,
+    *,
+    run_types: Iterable[sciline.typing.Key] | None,
+    monitor_types: Iterable[sciline.typing.Key] | None,
+) -> None:
+    # Remove all nodes that use a run type or monitor types that is
+    # not listed in the function arguments.
+    excluded_run_types = _excluded_type_args(RunType, run_types)
+    excluded_monitor_types = _excluded_type_args(MonitorType, monitor_types)
+    excluded_types = excluded_run_types | excluded_monitor_types
+
+    graph = workflow.underlying_graph
+    to_remove = [
+        node for node in graph if excluded_types & set(getattr(node, "__args__", set()))
```

Comment:
Does this work with nesting? One example to consider is the ESSsans `TransmissionRun`, e.g., in `NeXusComponent[IncidentMonitor, TransmissionRun[SampleRun]]`. But there can also be other types of nesting, I think?

Reply:
Works in SANS: https://github.com/scipp/esssans/pull/190

Reply:
So we will assume that no other nesting is needed? That is, we will not support, say `dict[str, Filename[RunType]]`?

Reply:
In this concrete example it is ok. And I am not aware of others.

We can also postpone nesting support until later, if required?

Reply:
Yes, I would say so